### ISSUE TITLE
a:where(:not(.wp-element-button)) is overwriting all anchor tags on site in 14.0.3

### ISSUE THREAD

**Issue by cb911nr** [WordPress/gutenberg#44166]
### Description

Installed the latest plugin for development features, and it's adding new CSS for 

```
a:where(:not(.wp-element-button)) {
                text-decoration: underline;
            }
```

Inside  <style id='global-styles-inline-css' type='text/css'>

But this is adding text-decoration underline to all anchors on our site, because this loads after our stylesheet..

Is there a way to remove this css line? Or have another way of dealing with it like getting this to inject prior to our stylesheet? 

### Step-by-step reproduction instructions

Install the 14.0.3 release plugin, styles are then outputted

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

**Comment by cb911nr** [comment]
Was able to just remove line 95 from

[lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php](https://github.com/WordPress/gutenberg/blob/48e841b70a08d77a4ea550c0d4314c3eb4418fa8/lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php)

Which resolved this issue.

**Comment by straightvisions-matthias-bathke** [comment]
You shouldn't remove or edit lines in core code, as it's not update save. Changing this default behavior is theme territory.

Disable **default link text decoration underline** _injected by Gutenberg_ via theme.json:
`{
   "version":2,
   "styles":{
      "elements":{
         "link":{
            "typography":{
               "textDecoration":false
            }
         }
      }
   }
}`

output of `a:where(:not(.wp-element-button)) { text-decoration: underline; }` will be omitted then.

**Comment by ddegner** [comment]
Still suffering from the same bug

**Comment by FabianKielmann** [comment]
> You shouldn't remove or edit lines in core code, as it's not update save. Changing this default behavior is theme territory.
> 
> Disable **default link text decoration underline** _injected by Gutenberg_ via theme.json: `{ "version":2, "styles":{ "elements":{ "link":{ "typography":{ "textDecoration":false } } } } }`
> 
> output of `a:where(:not(.wp-element-button)) { text-decoration: underline; }` will be omitted then.

That sadly removes the underline for any paragraph text that is given a hyperlink in the gutenberg editor :(

The text decoration mentioned here adds underlines for many of our elements in the customizer preview of our website, which sucks.

Removing this via theme.json is not an option though, as mentioned above.

**Comment by tom-warham** [comment]
Wether this is default functionality or not I can confirm this seems more like a bug than a feature. 

In a real world example just discovered, The `a:where(:not(.wp-element-button))` generated by the sitewide "link" styling takes priority over all a tag links sitewide. When trying to add a cover block inside of a query loop I would like the linked "title" block to be a larger font size in the query loop, if the title block settings "Make title a link" is selected then the title styles have no effect over the above `a:where(:not(.wp-element-button))` css rule. 

I would expect default functionality to be sitewide styles take least priority with individual block settings overriding the sitewide styling. 

I will try and create a "Styles" rule as above to resolve this issue but its very strange functioanlity as it stands. 

Any help apreciated TIA.

**Comment by Tirjasdyn** [comment]
I'm having a similar issue and a style rule is not working. I had to change the default Secondary color.

**Comment by yoche2001** [comment]
One workaround is to undo the change by using 
```
:root :where(a:where(:not(.wp-element-button))) {
  text-decoration: inherit;
}

```

**Comment by chillwebdesigns** [comment]
This worked for me in the theme.json

```
{
  "$schema": "http://schemas.wp.org/trunk/theme.json",
  "version": 3,
  "styles": {
    "elements": {
      "link": {
        "typography": {
          "textDecoration": "none"
        }
      }
    }
  }
}

```

**Comment by abhi36** [comment]
@chillwebdesigns solution actually worked. But it sets an invalid value for the `text-decoration` property. It has to be `none` in place of `false`. And the suggested version is `3` instead of `2`.

```json
{
  "$schema": "http://schemas.wp.org/trunk/theme.json",
  "version": 3,
  "styles": {
    "elements": {
      "link": {
        "typography": {
          "textDecoration": "none"
        }
      }
    }
  }
}
```

**Comment by Shoujeng** [comment]
The solution of @abhi36 works, but it also make the ```a:hover {textDecoration: underline}``` invalid. Hope it could be removed.

**Comment by tom-warham** [comment]
Hey,

 I'm usually away from keyboard on Fridays.

 I periodically check emails, but if you require assistance please
contact ***@***.*** or ***@***.***

 Cheers, Tom

**Comment by abhi36** [comment]
@Shoujeng you could extend the json just a bit to ensure explicit support on hover customisation.

you could put the following inside "link" alongside what's already there.
```json
":hover": {
      "typography": {
        "textDecoration": "none"
      }
    }
```
**

final json part should be

```json
{
  "$schema": "http://schemas.wp.org/trunk/theme.json",
  "version": 3,
  "styles": {
    "elements": {
      "link": {
        ":hover": {
          "typography": {
            "textDecoration": "none"
          }
        },
        "typography": {
          "textDecoration": "none"
        }
      }
    }
  }
}
```

**Comment by Shoujeng** [comment]
Thanks @abhi36 , a good temporal solution👍

**Comment by t-hamano** [comment]
I would like to understand what is considered a bug in this issue.

As already suggested, you should be free to change the underline behavior via theme.json.

**Comment by djmtype** [comment]
Unfortunately, none of these solutions are perfect. There should be another option for `false`, period.

I prefer `:where(a)`, while it abides my rules for `:hover`, it doesn't for `:focus-visible` due to specificity: 

```json
  "styles": {
    "elements": {
      "link": {
        ":hover": {
          "typography": {
            "textDecoration": "none"
          }
        },
        ":focus-visible": {
          "typography": {
            "textDecoration": "none"
          }
        },
        "typography": {
          "textDecoration": "underline"
        }
      }
    }
  },
```


The following doesn't work because I have it defined within a `@layer`: 

```
a:where(:not(.wp-element-button)) {
  &:focus-visible {
    text-decoration: none;
  }
 }
```

**Comment by t-hamano** [comment]
FWIW, there is a way to disable `a:where(:not(.wp-element-button))` style:

```php
/**
 * Custom hook to remove the following inline style
 *
 * a:where(:not(.wp-element-button)){text-decoration: underline;}
 */
function filter_theme_json_theme( $theme_json ){
	$theme_json_data = $theme_json->get_data();
	if ( isset( $theme_json_data['styles']['elements']['link']['typography']['textDecoration'] ) ) {
		unset( $theme_json_data['styles']['elements']['link']['typography']['textDecoration'] );
	}
	return new WP_Theme_JSON_Data( $theme_json_data, 'default' );
}
add_filter( 'wp_theme_json_data_default', 'filter_theme_json_theme' );
```

**Comment by tom-warham** [comment]
Hey,

 I'm usually away from keyboard on Fridays.

 I periodically check emails, but if you require assistance please
contact ***@***.*** or ***@***.***

 Cheers, Tom

**Comment by djmtype** [comment]
> FWIW, there is a way to disable `a:where(:not(.wp-element-button))` style:
> 
> /**
>  * Custom hook to remove the following inline style
>  *
>  * a:where(:not(.wp-element-button)){text-decoration: underline;}
>  */
> function filter_theme_json_theme( $theme_json ){
> 	$theme_json_data = $theme_json->get_data();
> 	if ( isset( $theme_json_data['styles']['elements']['link']['typography']['textDecoration'] ) ) {
> 		unset( $theme_json_data['styles']['elements']['link']['typography']['textDecoration'] );
> 	}
> 	return new WP_Theme_JSON_Data( $theme_json_data, 'default' );
> }
> add_filter( 'wp_theme_json_data_default', 'filter_theme_json_theme' );

@t-hamano Thanks! By that logic, in theme.json I've added "default". It could've also been "foobar" since "default" will return invalid. Then, it falls back to next valid value in the cascade. It's a good enough hack, but `false` is really needed so there's no invalid CSS to boot, and `a:where(:not(.wp-element-button))` is never returned. 

BTW, The results are the same if it's declared in theme.json or via filter above.

```json
"styles": {
    "elements": {
      "link": {
        "typography": {
          "textDecoration": "default"
        }
      }
    }
  }
```

![Image](https://github.com/user-attachments/assets/55982d89-e431-4437-8c6b-99e9a046e4a5)

**Comment by t-hamano** [comment]
@djmtype Does my proposed approach solve all your problems? If not, can you tell me what you want?

**Comment by djmtype** [comment]
@t-hamano while it satisfies my needs, it's more like a hack than a solution. You can see what happens based on my screenshot. The "default" value is invalid, so, it falls back to the style I had defined for the link element with lower specificity. The ideal solution is to be able to filter out Wordpress' class name from the start. They allow `false` on other Gutenberg properties. Unsure why it's not allowed here. 

In other words, since we never asked for this, `a:where(:not(.wp-element-button))`, we should have the option to remove it.

**Comment by t-hamano** [comment]
> `a:where(:not(.wp-element-button))`, we should have the option to remove it. 

Can't you achieve that with the code I suggested in [this comment](https://github.com/WordPress/gutenberg/issues/44166#issuecomment-2760647337)?

**Comment by djmtype** [comment]
@t-hamano unless I had caching issues, your filter adds "default" as a value to a:where(:not(.wp-element-button)). It didn't remove the class name altogether. 

Your filter yielded the same results as this code, in theme.json. 

```json
"styles": {
    "elements": {
      "link": {
        "typography": {
          "textDecoration": "default"
        }
      }
    }
  }
```
I would like the text decoration value to accept false, so it never returns that classname.

**Comment by t-hamano** [comment]
@djmtype 

Are you talking about "default" in this code?

```php
return new WP_Theme_JSON_Data( $theme_json_data, 'default' );
```

This "defaults" represents the origin of theme.json, not a CSS value. Please check the code I provided again.

**Comment by djmtype** [comment]
@t-hamano It was stale cache on my end. The Sage theme seems to do that. 
Your filter seems to be the solution! 

There should be, and maybe there is, a theme.json equivalent? 

The following seems to work, too, with theme.json. Whether it's valid is another story. This is different than what I posted earlier BTW. 

```json
"styles": {
    "elements": {
      "link": {
        "typography": {
          "textDecoration": {
            "ref": "default"
          }
        }
      }
    }
  },
```

**Comment by t-hamano** [comment]
@djmtype Can we close this issue?

**Comment by djmtype** [comment]
@t-hamano Your filter is the answer for me. I don't know if there are any other edge cases.

**Comment by t-hamano** [comment]
Ok, so let's close this for now. If there are any problem I've missed, feel free comment.

**Comment by djmtype** [comment]
@t-hamano I should probably open a new issue. I have the same problem for :where(figure). Wordpress automatically adds margin: 0 0 1rem. 

I can't override this because my design system uses  @layer, so my styles have a lower specificity. 

Have you come across this?

**Comment by t-hamano** [comment]
> I can't override this because my design system uses https://github.com/layer, so my styles have a lower specificity.

Regarding the conflicts between WordPress's design system and the custom design that uses `@layer`, some issues have been reported:

- #69833
- #51128

Are these issues useful to you?

**Comment by HowdyMcGee** [comment]
Why is this being applied to themes that **are not** Block Based themes? I would expect those to be excluded from global inline styles such as this on non-block themes.

**Comment by t-hamano** [comment]
@HowdyMcGee 

> Why is this being applied to themes that **are not** Block Based themes?

In what environment have you verified this? As far as I've tested, the classic theme does not have the underline style for a elements by default.
